compaction/compaction_manager.hh

```diff
@@ -475,6 +475,7 @@ protected:
     sstables::compaction_data _compaction_data;
     state _state = state::none;
     throw_if_stopping _do_throw_if_stopping;
+    sstables::compaction_read_monitor_generator_wrapper _progress_monitor;
 
```

Comment:
I liked the idea of the wrapper, so not to expose details and to handle the problem in which the monitor may behave differently based on the compaction type

Reply:
An alternative approach, so we won't have to do dynamic casts and have a wrapper, is to declare compaction_read_monitor_generator in compaction.hh. Then compaction_task_executor can store  `std::unique_ptr<sstables::compaction_read_monitor_generator>` and we pass sstables::compaction_read_monitor_generator as a reference to compact_sstables(). We set up a new generator on compaction_task_executor::setup_new_compaction(...) and reset it in compaction_task_executor::finish_compaction(). The knowledge of whether monitor needs to update backlog tracker belongs to each executor of every compaction type, which is essentially the same as today.

Reply:
If you don't mind I would stay with the wrapper.

We can't keep `sstables::compaction_read_monitor_generator` in executors for their whole lifetime since `compaction_read_monitor`'s destructor updates backlog tracker. Thus we would need to keep a variable which we would use to get progress when generator is set to nullptr. That's similar to what's done in wrapper but I think it's clearer to extract it.

It also makes it easier to handle uncommon way of counting progress and we will probably need it for regular compaction. Please see #15171. Even if it won't stay as is I believe we will need to do it different way than it's done here.

Reply:
Yes, I think it's also less fragile as the generator lifetime is bounded to compaction's lifetime, and doesn't depend on executor creating and clearing on either success or failure.